src/main/java/core/basesyntax/Car.java

```diff
@@ -1,13 +1,87 @@
 package core.basesyntax;
 
+import java.util.ArrayList;
 import java.util.List;
+import java.util.Objects;
 
-/**
- * Make this class immutable. See requirements in the README file
- */
-public class Car {
-    private int year;
-    private String color;
-    private List<Wheel> wheels;
-    private Engine engine;
+public final class Car {
+    private final int year;
+    private final String color;
+    private final List<Wheel> wheels;
+    private final Engine engine;
+
+    public Car(int year, String color, List<Wheel> wheels, Engine engine) {
+        this.year = year;
+        this.color = color;
+        this.wheels = cloneWheels(wheels);
+        this.engine = engine != null ? engine.clone() : null;
+    }
+
+    public Car addWheel(Wheel wheel) {
+        List<Wheel> newWheels = new ArrayList<>(wheels);
+        newWheels.add(wheel);
+        return new Car(year, color, newWheels, engine);
+    }
+
+    public Car changeEngine(Engine engine) {
+        return new Car(year, color, wheels, engine);
+    }
+
+    public Car changeColor(String blue) {
+        return new Car(year, blue, wheels, engine);
+    }
+
+    public int getYear() {
+        return year;
+    }
+
+    public String getColor() {
+        return color;
+    }
+
+    public List<Wheel> getWheels() {
+        return new ArrayList<>(wheels);
+    }
+
+    public Engine getEngine() {
+        return engine != null ? engine.clone() : null;
+    }
+
+    private List<Wheel> cloneWheels(List<Wheel> wheels) {
+        List<Wheel> wheelsClones = new ArrayList<>();
+        for (Wheel wheel : wheels) {
+            wheelsClones.add(wheel.clone());
+        }
+        return wheelsClones;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        Car car = (Car) o;
+        if (year != car.year) {
+            return false;
+        }
+        if (!Objects.equals(color, car.color)) {
+            return false;
+        }
+        if (!Objects.equals(wheels, car.wheels)) {
+            return false;
+        }
+        return Objects.equals(engine, car.engine);
```

Comment:
```suggestion
        return year == car.year && Objects.equals(color, car.color)
                && Objects.equals(wheels, car.wheels) && Objects.equals(engine, car.engine);
```
We can simplify our code and reject redundant  if